core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestRewriteCaseExpressionPredicate.java

```diff
@@ -0,0 +1,266 @@
+import static io.trino.SystemSessionProperties.OPTIMIZE_CASE_EXPRESSION_PREDICATE;
+import static io.trino.metadata.MetadataManager.createTestMetadataManager;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static io.trino.sql.planner.TestingPlannerContext.plannerContextBuilder;
+import static io.trino.sql.planner.TypeAnalyzer.createTestingTypeAnalyzer;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.filter;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.values;
+import static io.trino.sql.planner.iterative.rule.SimplifyExpressions.rewrite;
+import static io.trino.transaction.TransactionBuilder.transaction;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestRewriteCaseExpressionPredicate
+        extends BaseRuleTest
+{
+    private static final Metadata METADATA = createTestMetadataManager();
+    private static final PlannerContext PLANNER_CONTEXT = plannerContextBuilder().withMetadata(METADATA).build();
+    private static final DomainTranslator DOMAIN_TRANSLATOR = new DomainTranslator(PLANNER_CONTEXT);
+    private static final Map<Symbol, Type> TYPE_MAP = ImmutableMap.of(new Symbol("col1"), INTEGER, new Symbol("col2"), INTEGER, new Symbol("col3"), VARCHAR);
+
+    private void assertRewriteDoesNotFire(String expression)
+    {
+        tester().assertThat(new RewriteCaseExpressionPredicate(tester().getPlannerContext(), tester().getTypeAnalyzer()))
+                .setSystemProperty(OPTIMIZE_CASE_EXPRESSION_PREDICATE, "true")
+                .on(p -> p.filter(PlanBuilder.expression(expression), p.values()))
+                .doesNotFire();
+    }
+
+    private void assertRewrittenAndOptimizedExpression(String inputExpressionStr,
+                                                       String expectedRewrittenExpressionStr,
+                                                       String expectedOptimizedExpressionStr)
+    {
+        Expression inputExpression = PlanBuilder.expression(inputExpressionStr);
+        Expression expectedRewrittenExpression = PlanBuilder.expression(expectedRewrittenExpressionStr);
+
+        tester().assertThat(new RewriteCaseExpressionPredicate(tester().getPlannerContext(), tester().getTypeAnalyzer()))
+                .setSystemProperty(OPTIMIZE_CASE_EXPRESSION_PREDICATE, "true")
+                .on(p -> p.filter(inputExpression, p.values(p.symbol("col1"), p.symbol("col2"), p.symbol("col3"))))
+                .matches(filter(expectedRewrittenExpression, values("col1", "col2", "col3")));
+
+        Expression expectedOptimizedExpression = PlanBuilder.expression(expectedOptimizedExpressionStr);
+
+        Expression optimizedExpression = getOptimizedExpression(expectedRewrittenExpression);
+        assertThat(optimizedExpression).isEqualTo(expectedOptimizedExpression);
+    }
+
+    private Expression getOptimizedExpression(Expression expression)
```

Comment:
In fact, a unit test for the rule should be limited to what the rule does. So, the "getOptimized" part does not belong here. Moreover, this optimization does not directly reflect what happens in Trino's optimization process.

If you want to test the final form of the expression, you can do this in `TestLogicalPlanner`.

Reply:
Done.